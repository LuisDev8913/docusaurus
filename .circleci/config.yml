aliases:
  - &root-yarn |
    yarn install --non-interactive --cache-folder ~/.cache/yarn

  - &root-restore-yarn-cache
    keys:
      - root-yarn-{{ .Branch }}-{{ checksum "yarn.lock" }}
      # Fallback in case checksum fails
      - root-yarn-{{ .Branch }}-

  - &root-save-yarn-cache
    paths:
      - node_modules
      - ~/.cache/yarn
    key: root-yarn-{{ .Branch }}-{{ checksum "yarn.lock" }}

  - &filter-only-master
    branches:
      only:
        - master

  - &filter-ignore-gh-pages
    branches:
      ignore: gh-pages

defaults: &defaults
  working_directory: ~/docusaurus
  docker:
    - image: circleci/node:12
  resource_class: large

version: 2
jobs:
  lint-prettier:
    <<: *defaults
    steps:
      - checkout
      - restore-cache: *root-restore-yarn-cache
      - run: *root-yarn
      - save-cache: *root-save-yarn-cache
      - run:
          name: Check ESLint
          command: yarn lint
      - run:
          name: Check Prettier
          command: yarn prettier:diff

  tests:
    <<: *defaults
    steps:
      - checkout
      - restore-cache: *root-restore-yarn-cache
      - run: *root-yarn
      - save-cache: *root-save-yarn-cache
      - run:
          name: Run Test Suites
          command: yarn test

  # The CIRCLE_ variables are defined during the CircleCI build process
  # https://circleci.com/docs/1.0/environment-variables/
  deploy-website:
    <<: *defaults
    steps:
      - checkout
      - restore-cache: *root-restore-yarn-cache
      - run: *root-yarn
      - save-cache: *root-save-yarn-cache
      - run:
          name: Configure GitHub Bot
          # Do not do this if we don't have the right org (facebook), or if this is just a pull request
          command: |
            if [[ $CIRCLE_PROJECT_USERNAME == "facebook" && -z $CIRCLE_PR_USERNAME ]]; then
              git config --global user.email "docusaurus-bot@users.noreply.github.com"
              git config --global user.name "Website Deployment Script"
              echo "machine github.com login docusaurus-bot password $DOCUSAURUS_PUBLISH_TOKEN" > ~/.netrc
            fi
      - run:
          name: Deploy v1 Website
          # Skip the deploy if we don't have the right org (facebook), or if this is just a pull request
          command: |
            if ! git diff-tree --no-commit-id --name-only -r HEAD | grep -E "(docusaurus-1\.x\/.*)|(website-1\.x\/.*)"; then
              echo "Skipping deploy. No relevant v1 website files have changed"
            elif [[ $CIRCLE_PROJECT_USERNAME == "facebook" && -z $CIRCLE_PR_USERNAME ]]; then
              echo "Deploying website v1..."
              cd website-1.x

              # install Docusaurus and generate file of English strings
              yarn run write-translations

              # upload translation strings and download translations
              # TODO we have a weird Crowdin error "Project already contains the file", trying to resolve that with support
              # In the meantime we should not block the v1 site deployment
              yarn crowdin-upload || echo "Crowdin source upload failure, but we'll continue..."

              # download only enabled languages
              yarn crowdin-download
              # for lang in fr ko ru ro pt-BR zh-CN
              # do
              #    yarn crowdin-download -l $lang
              # done

              # publish
              GIT_USER=docusaurus-bot USE_SSH=false FORCE_DEPLOY=true yarn run publish-gh-pages
            else
              echo "Skipping deploy."
            fi

# Workflows enables us to run multiple jobs in parallel
workflows:
  version: 2

  lint-prettier:
    jobs:
      - lint-prettier:
          filters: *filter-ignore-gh-pages

  tests-and-deploy:
    jobs:
      - tests:
          filters: *filter-ignore-gh-pages
      - deploy-website:
          requires:
            - tests
          filters: *filter-only-master
