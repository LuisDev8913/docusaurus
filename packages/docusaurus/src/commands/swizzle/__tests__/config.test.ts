/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {SwizzleConfig} from '@docusaurus/types';
import {normalizeSwizzleConfig} from '../config';

describe('normalizeSwizzleConfig', () => {
  test(`validate no components config`, async () => {
    const config: SwizzleConfig = {
      components: {},
    };
    expect(normalizeSwizzleConfig(config)).toEqual(config);
  });

  test(`validate complete config`, async () => {
    const config: SwizzleConfig = {
      components: {
        SomeComponent: {
          actions: {
            wrap: 'safe',
            eject: 'unsafe',
          },
          description: 'SomeComponent description',
        },
        'Other/Component': {
          actions: {
            wrap: 'forbidden',
            eject: 'unsafe',
          },
          description: 'Other/Component description',
        },
      },
    };
    expect(normalizeSwizzleConfig(config)).toEqual(config);
  });

  test(`normalize partial config`, async () => {
    const config: SwizzleConfig = {
      components: {
        SomeComponent: {
          // @ts-expect-error: incomplete actions map
          actions: {
            eject: 'safe',
          },
          description: 'SomeComponent description',
        },
        'Other/Component': {
          // @ts-expect-error: incomplete actions map
          actions: {
            wrap: 'forbidden',
          },
        },
      },
    };
    expect(normalizeSwizzleConfig(config)).toMatchInlineSnapshot(`
      Object {
        "components": Object {
          "Other/Component": Object {
            "actions": Object {
              "eject": "unsafe",
              "wrap": "forbidden",
            },
          },
          "SomeComponent": Object {
            "actions": Object {
              "eject": "safe",
              "wrap": "unsafe",
            },
            "description": "SomeComponent description",
          },
        },
      }
    `);
  });

  test(`reject missing components`, async () => {
    // @ts-expect-error: incomplete actions map
    const config: SwizzleConfig = {};

    expect(() =>
      normalizeSwizzleConfig(config),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Swizzle config does not match expected schema: \\"components\\" is required"`,
    );
  });

  test(`reject invalid action name`, async () => {
    const config: SwizzleConfig = {
      components: {
        MyComponent: {
          actions: {
            wrap: 'safe',
            eject: 'unsafe',
            // @ts-expect-error: on purpose
            bad: 'safe',
          },
        },
      },
    };

    expect(() =>
      normalizeSwizzleConfig(config),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Swizzle config does not match expected schema: \\"components.MyComponent.actions.bad\\" is not allowed"`,
    );
  });

  test(`reject invalid action status`, async () => {
    const config: SwizzleConfig = {
      components: {
        MyComponent: {
          actions: {
            wrap: 'safe',
            // @ts-expect-error: on purpose
            eject: 'invalid-status',
          },
        },
      },
    };

    expect(() =>
      normalizeSwizzleConfig(config),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Swizzle config does not match expected schema: \\"components.MyComponent.actions.eject\\" must be one of [safe, unsafe, forbidden]"`,
    );
  });
});
