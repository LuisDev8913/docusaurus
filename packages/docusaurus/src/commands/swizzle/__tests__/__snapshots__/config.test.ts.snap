// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeSwizzleConfig normalize partial config 1`] = `
Object {
  "components": Object {
    "Other/Component": Object {
      "actions": Object {
        "eject": "unsafe",
        "wrap": "forbidden",
      },
    },
    "SomeComponent": Object {
      "actions": Object {
        "eject": "safe",
        "wrap": "unsafe",
      },
      "description": "SomeComponent description",
    },
  },
}
`;
